phoenix-queryserver/src/main/java/org/apache/phoenix/queryserver/server/QueryServer.java

```diff
@@ -233,16 +240,29 @@ public int run(String[] args) throws Exception {
       // Build and start the HttpServer
       server = builder.build();
       server.start();
+      registerToServiceProvider(hostname);
       runningLatch.countDown();
       server.join();
       return 0;
     } catch (Throwable t) {
       LOG.fatal("Unrecoverable service error. Shutting down.", t);
       this.t = t;
       return -1;
+    } finally {
+      deRegister();
     }
   }
 
+  private void registerToServiceProvider(String hostName) throws Exception {
+      PqsZookeeperConf pqsZookeeperConf = new PqsZookeeperConfImpl(getConf());
```

Comment:
Oh, I see now. You're instantiating these implementations directly.

I think you should use Java's ServiceLoader to properly separate properly separate interface from implementation across the queryserver and loadbalancer modules. Here in queryserver, you would just refer to PqsZookeeperConf and Registry (but the concrete implementations would be wired up by ServiceLoader as defined by the META-INF/services/... files you create). 

If you're not familiar: https://docs.oracle.com/javase/tutorial/ext/basics/spi.html. Should be pretty straightforward. e.g. this is also how the PhoenixDriver class is loaded for JDBC (so there's example of what to do in phoenix-core).